guest-owner-proxy/guest-owner-proxy.py

```diff
@@ -0,0 +1,254 @@
+from concurrent import futures
+from os import path, mkdir, makedirs, urandom
+from uuid import UUID
+from Crypto.Cipher import AES
+from Crypto.Util import Counter
+
+from pre_attestation_pb2 import BundleResponse, SecretResponse
+import pre_attestation_pb2_grpc
+
+import grpc
+import json
+import subprocess
+import time
+import hmac
+import base64
+import hashlib
+
+keysets = {}
+sevtool_path = "/home/tobin/sev-tool/src/sevtool"
+certs_path = "/tmp/sev-guest-owner-proxy/certs/"
+connection_id = 0
+
+def guid_to_le(guid_str):
+    return UUID("{" + guid_str + "}").bytes_le
+
+def construct_secret_entry(guid_str, secret_bytes):
+    l = 16 + 4 + len(secret_bytes)
+    entry = bytearray(l);
+    entry[0:16] = guid_to_le(guid_str)
+    entry[16:20] = l.to_bytes(4, byteorder='little')
+    entry[20:20+len(secret_bytes)] = secret_bytes
+    return entry
+
+
+def cli_parsing():
+    usage = '''usage: %prog [options] [command]
+    '''
+
+    _parser = OptionParser(usage)
+
+    _parser.add_option("-p", "--port", \
+                       dest = "grpc_port", \
+                       default = 50051, \
+                       help = "Set port for gRPC server.")
+
+    _parser.add_option("-c", "--config", \
+                       dest = "config_path", \
+                       default = "keysets.json", \
+                       help = "Path to keyset config file.")
+
+    _parser.add_option("-k", "--keyfile", \
+                       dest = "keyfile_path", \
+                       default = "keys.json", \
+                       help = "Path to keyfile.")
+
+    _parser.set_defaults()
+    (_options, _args) = _parser.parse_args()
+
+    return _options
+
+def main(options):
+    try:
+        with open(options.config_path) as f:
+            global keysets
+            keysets = json.load(f)
+            print(keysets)
+    except Exception as e:
+        print("Failed to load config: {}".format(e.msg))
+
+    # this will be replaced by an HSM probably
+    try:
+        with open(options.keyfile_path) as f:
+            global keys
+            keys = json.load(f)
+    except Exception as e:
+        print("Failed to load keyfile: {}".format(e.msg))
+
+
+    #makedirs(certs_path, exist_ok = True)
+
+    # this demo implementation isn't designed for parallelism.
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=1))
+    pre_attestation_pb2_grpc.add_SetupServicer_to_server(SetupService(), server)
+
+    server.add_insecure_port("[::]:{}".format(options.grpc_port))
+    server.start()
+    server.wait_for_termination()
+
+
+class SetupService(pre_attestation_pb2_grpc.SetupServicer):
+    def __init__(self):
+        self.connection_id = 1
+
+    def GetLaunchBundle(self, request, context):
+        print("Servicing Launch Bundle Request")
+
+        # make dir for this conection
+        connection_certs_path = path.join(certs_path, "connection{}".format(self.connection_id))
+        makedirs(connection_certs_path, exist_ok=True)
```

Comment:
Here you create dir `<certs_path>/connection0`, and later increment self.connection_id and return it to the client. So the response will have field ConnectionId=1 but the directory created here will be `.../connection0`.

I think the increment of self.connection_id should be the first thing in this function.

Reply:
whoops